contracts/OtokenFactory.sol

```diff
@@ -0,0 +1,185 @@
+pragma solidity =0.6.10;
+
+import {OtokenSpawner} from "./OtokenSpawner.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
+import {OtokenInterface} from "./interfaces/OtokenInterface.sol";
+import {WhitelistInterface} from "./interfaces/WhitelistInterface.sol";
+
+/**
+ * SPDX-License-Identifier: UNLICENSED
+ * @title A factory for opyn tokens
+ * @author Opyn
+ * @notice Create new otokens and keep track of all created tokens.
+ * @dev Calculate contract address before each creation with CREATE2
+ * and deploy eip-1167 minimal proxies for otoken logic contract.
+ */
+contract OtokenFactory is OtokenSpawner {
+    /// @notice The Opyn AddressBook contract that records addresses of whitelist module and otoken impl address. */
+    address public addressBook;
+
+    /// @dev An array of all created otokens */
+    address[] private _otokens;
+
+    /// @dev A mapping from parameters hash to its deployed address
+    mapping(bytes32 => address) private _idToAddress;
+
+    constructor(address _addressBook) public {
+        addressBook = _addressBook;
+    }
+
+    /* Event: OtokenCreated - Emitted when factory create a new Option. */
+    event OtokenCreated(
+        address indexed tokenAddress,
+        address indexed creator,
+        address underlying,
+        address strike,
+        address collateral,
+        uint256 strikePrice,
+        uint256 expiry,
+        bool isPut
+    );
+
+    /**
+     * @notice create new otokens
+     * @dev deploy an eip-1167 minimal proxy with CREATE2 and register it to the whitelist module.
+     * @param _underlyingAsset asset that the option references
+     * @param _strikeAsset asset that the strike price is denominated in
+     * @param _collateralAsset asset that is held as collateral against short/written options
+     * @param _strikePrice strike price with decimals = 18
+     * @param _expiry expiration timestamp in second
+     * @param _isPut is this a put option, if not it is a call
+     * @return newOtoken address of the newly created option
+     */
+    function createOtoken(
+        address _underlyingAsset,
+        address _strikeAsset,
+        address _collateralAsset,
+        uint256 _strikePrice,
+        uint256 _expiry,
+        bool _isPut
+    ) external returns (address newOtoken) {
```

Comment:
do we need a return address?

Reply:
```suggestion
    ) external {
```

Reply:
I think it would be helpful in terms of contract integration. 

For example for the following flow:
```
address otoken = Factory.getOtoken(params...)
if(otoken = address(0) ) {
   otoken = Factory.createOtoken(params...)
} 
```

Reply:
I don't think that's a good idea, we should make that check in Factory.
In case `otoken == address(0)` there will be two calls to the Factory, which is more expensive in term of gas.

Reply:
hmm it breaks our syntax of not having returns in external functions, but if it is helpful in this case, that's ok imo. @haythem96 thoguths?